Framework/Algorithms/src/FindPeaksConvolve.cpp

```diff
@@ -0,0 +1,459 @@
+
+    extractPeaks(dataIndex, iOverSigConvOutput, xData, yData, kernelBinCount.first / 2);
+
+    if (m_createIntermediateWorkspaces) {
+      createIntermediateWorkspaces(dataIndex, kernel, iOverSigConvOutput, xData);
+    }
+  }
+}
+
+Tensor1D FindPeaksConvolve::createKernel(const int binCount) const {
+  Tensor1D kernel(binCount);
+  for (int i{0}; i < binCount; i++) {
+    // create integrating shoebox kernel with central positive region & negative background shell with ~the same
+    // elements in each.
+    if (i < std::ceil(binCount) * 0.25 || i >= binCount * 0.75) {
+      kernel.data()[i] = -1.0;
+    } else {
+      kernel.data()[i] = 1.0;
+    }
+  }
+  return kernel;
+}
+
+Tensor1D FindPeaksConvolve::createSmoothKernel(const size_t kernelSize) const {
+  Tensor1D kernel(kernelSize);
+  for (size_t i{0}; i < kernelSize; i++) {
+    kernel.data()[i] = 1.0 / static_cast<double>(kernelSize);
+  }
+  return kernel;
+}
+
+std::pair<size_t, bool> FindPeaksConvolve::getKernelBinCount(const HistogramData::HistogramX *xData) const {
+  const double peakExtent = getProperty("EstimatedPeakExtent");
+  const int peakExtentNBins = getProperty("EstimatedPeakExtentNBins");
+  bool sizeError{false};
+
+  size_t kernelBinCount{0};
+  if (peakExtent != EMPTY_DBL()) {
+    const double x1{xData->rawData()[static_cast<size_t>(std::floor((xData->size() - 1) / 2))]};
+    const double x2{xData->rawData()[static_cast<size_t>(std::floor((xData->size() - 1) / 2)) + 1]};
+    kernelBinCount = static_cast<size_t>(std::floor(peakExtent * 2 / (x2 - x1)));
+  } else {
+    kernelBinCount = static_cast<size_t>(peakExtentNBins);
+  }
+  if (kernelBinCount > xData->size()) {
+    sizeError = true;
+  }
+  return std::make_pair(kernelBinCount, sizeError);
+}
+
+Eigen::VectorXd FindPeaksConvolve::centreBinsXData(const HistogramData::HistogramX *xData) const {
+  Eigen::VectorXd xDataVec{0.5 * (EigenMap_const(&xData->front(), xData->size() - 1) +
+                                  EigenMap_const(&xData->front() + 1, xData->size() - 1))};
+  return xDataVec;
+}
+
+void FindPeaksConvolve::extractPeaks(const size_t dataIndex, const Tensor1D &iOverSigma,
+                                     const HistogramData::HistogramX *xData, const TensorMap_const &yData,
+                                     const size_t peakExtentBinNumber) {
+  int dataPointCount{0};
+  std::pair<int, double> dataRegionMax{0, 0.0};
+  std::vector<FindPeaksConvolve::PeakResult> peakCentres;
+  for (auto i{0}; i < iOverSigma.size(); i++) {
+    if (iOverSigma(i) > m_iOverSigmaThreshold) {
+      if (dataPointCount == 0) {
+        dataRegionMax = std::make_pair(i, iOverSigma(i));
+      } else {
+        if (iOverSigma.data()[i] > dataRegionMax.second) {
+          dataRegionMax = {i, iOverSigma(i)};
+        }
+      }
+      dataPointCount++;
+    } else if (iOverSigma(i) <= 0 || !m_mergeNearbyPeaks || i == iOverSigma.size() - 1) {
+      if (dataPointCount >= 2) {
```

Comment:
Is this hard-coding that the minimum size of peak must contain at least 2 bins over the I/sigma threshold? I'm not against this, it helps remove spurious peaks, but it is coupled to the I/sigma value chosen. At some point it would be nice to set a minimum extent between zero-crossings as we know this corresponds to roughly 2 stdev for a Gaussian peak.  As the peak extent chosen should be ~4-6 stdev we would expect genuine peaks to have a number of bins between zero crossings to be about ~1/3 of the peak extent.  That's might be a bit strict (peaks may not be Gaussian, some overlap etc.) so perhaps ~1/5 would be a good default? Would this be a pain to put in?

Reply:
Yes it is as you interpreted, this hardcodes the minimum peak size to 2 bins over the threshold. On the face of it, I think it should be fairly easy to increment `dataPointCount` for any `iOverSigma` value above 0, then when recording a peak discard those below 1/5 of the bin count extent.

Will have to think about how this will work if we're NOT merging peaks. In that case, you could have two peaks that occur in the same "zero crossing" region. At the moment in this case we log peaks and reset the region count as soon as iOverSigma drops below the threshold. If we really wanted to we could hold onto multiple peaks then log them all  at the next zero crossing if `zero crossing region range = peak number * 1/5 threshold` or something...

Let me know what you think.

Reply:
Can we add this to the to do list, I could do with getting this in and adding the number of peaks functionality and am a bit pushed for time!